File: benchmarks/column_ops.nim
import os
import strutils
import tables
import times
import strinterp
import typetraits
import future
import macros

import ../src/columns

import arraymancer

type
  Benchmark = int -> float


# template to simplify timed execution
template runTimed(body: untyped): float =
  GC_fullCollect()
  let t1 = epochTime()
  body
  let t2 = epochTime()
  let runtime = (t2 - t1) * 1000
  runtime

# -----------------------------------------------------------------------------
# Benchmarks
# -----------------------------------------------------------------------------

proc benchmarkZeros[T](N: int): float =
  let runTime = runTimed:
    let col {.used.} = newCol[T](N)
  runTime

proc benchmarkZerosArraymancer[T](N: int): float =
  let runTime = runTimed:
    let col {.used.} = zeros[T](N)
  runTime

proc benchmarkSum[T](N: int): float =
  let col = newCol[T](N)
  let runTime = runTimed:
    let mean {.used.} = col.sum()
  runTime

proc benchmarkSumArraymancer[T](N: int): float =
  let col = zeros[T](N)
  let runTime = runTimed:
    let mean {.used.} = col.sum()
  runTime

proc benchmarkMax[T](N: int): float =
  let col = newCol[T](N)
  let runTime = runTimed:
    let max {.used.} = col.max(T)
  runTime

proc benchmarkMaxArraymancer[T](N: int): float =
  let col = zeros[T](N)
  let runTime = runTimed:
    let max {.used.} = col.max()
  runTime

# -----------------------------------------------------------------------------
# Runner helpers
# -----------------------------------------------------------------------------

proc runBenchmarkRepeated(benchmark: Benchmark, label: string, N: int, iterations: int) =
  var runTimeSum = 0.0
  var runTimeMin = Inf
  var runTimeMax = NegInf
  for i in 0 ..< iterations:
    let runTime = benchmark(N)
    runTimeSum += runTime
    if runTime < runTimeMin:
      runTimeMin = runTime
    if runTime > runTimeMax:
      runTimeMax = runTime
  let avgRunTime = runTimeSum / iterations.float
  # echo label, ": ", runtime
  echo fmt"${label}%-60s ${runTimeMin}%6.3f ms    ${avgRunTime}%6.3f ms    ${runTimeMax}%6.3f ms"

macro makeArrayOverTypes(name: untyped): untyped =
  ## Takes a generic function f and returns an array with generic instantiations:
  ## [(f[int16], "int16", (f[int32], "int32"), ...]
  result = newNimNode(nnkBracket)
  let typeSymbols = [
    bindsym"int16", bindsym"int32", bindsym"int64", bindsym"float32", bindsym"float64"
  ]
  for typeName in typeSymbols:
    result.add(newPar(
      newNimNode(nnkBracketExpr).add(name, typeName),
      newStrLitNode($typeName)
    ))
  # echo "result: ", result.repr

proc allBenchmarkZeros(N: int) =
  echo " *** Benchmark: zeros"
  for benchmark, label in makeArrayOverTypes(benchmarkZeros).items():
    runBenchmarkRepeated(benchmark, label, N, 100)
  echo " *** Benchmark: zeros (arraymancer)"
  for benchmark, label in makeArrayOverTypes(benchmarkZerosArraymancer).items():
    runBenchmarkRepeated(benchmark, label, N, 100)

proc allBenchmarkSum(N: int) =
  echo " *** Benchmark: sum"
  for benchmark, label in makeArrayOverTypes(benchmarkSum).items():
    runBenchmarkRepeated(benchmark, label, N, 100)
  echo " *** Benchmark: sum (arraymancer)"
  for benchmark, label in makeArrayOverTypes(benchmarkSumArraymancer).items():
    runBenchmarkRepeated(benchmark, label, N, 100)

proc allBenchmarkMax(N: int) =
  echo " *** Benchmark: max"
  for benchmark, label in makeArrayOverTypes(benchmarkMax).items():
    runBenchmarkRepeated(benchmark, label, N, 100)
  echo " *** Benchmark: max (arraymancer)"
  for benchmark, label in makeArrayOverTypes(benchmarkMaxArraymancer).items():
    runBenchmarkRepeated(benchmark, label, N, 100)


proc main() =
  let args = commandLineParams()
  if args.len == 0:
    echo "ERROR: Expected argument N."
    quit(1)
  let N = args[0].parseInt
  let selection = if args.len > 1: args[1] else: nil

  if selection.isNil or selection == "zeros":
    allBenchmarkZeros(N)
  if selection.isNil or selection == "sum":
    allBenchmarkSum(N)
  if selection.isNil or selection == "max":
    allBenchmarkMax(N)



when isMainModule:

  main()

File: benchmarks/column_ops.py
#!/usr/bin/env python

from __future__ import division, print_function

import numpy as np
import pandas as pd
import sys
import time
import gc


class TimedContext(object):

    def __init__(self):
        self.runtime = None

    def __enter__(self):
        gc.collect()
        self.t1 = time.time()
        return self

    def __exit__(self, exc_type, exc_val, exc_tb):
        self.t2 = time.time()
        self.runtime = (self.t2 - self.t1) * 1000
        # print("{:<60s} {:6.3f} ms".format(self.label, runtime))


def bench_zeros(dtype, N):
    with TimedContext() as ctx:
        data = np.zeros(N, dtype=dtype)
    assert len(data) == N
    return ctx.runtime


def bench_ones(dtype, N):
    with TimedContext() as ctx:
        data = np.ones(N, dtype=dtype)
    assert len(data) == N
    return ctx.runtime


def bench_range(dtype, N):
    with TimedContext() as ctx:
        data = np.arange(0, N, dtype=dtype)
    assert len(data) == N
    return ctx.runtime


def bench_sum(dtype, N):
    data = np.zeros(N, dtype)
    with TimedContext() as ctx:
        sum_val = data.sum()
    assert sum_val > -1
    return ctx.runtime


def bench_max(dtype, N):
    data = np.zeros(N, dtype)
    with TimedContext() as ctx:
        max_val = data.max()
    assert max_val > -1
    return ctx.runtime


def run_benchmark_repeated(benchmark, label, N, iterations, dtypes, result_file):
    print(" *** Benchmark: {}".format(label))
    for dtype in dtypes:
        runtime_sum = 0.0
        runtime_min = None
        runtime_max = None
        all_runtimes = []
        for i in xrange(iterations):
            runtime = benchmark(dtype, N)
            all_runtimes.append(runtime)
            runtime_sum += runtime
            if runtime < runtime_min or runtime_min is None:
                runtime_min = runtime
            if runtime > runtime_max or runtime_max is None:
                runtime_max = runtime
        runtime_avg = runtime_sum / iterations
        # echo label, ": ", runtime
        print(
            "{:<60s} {:6.3f} ms    {:6.3f} ms    {:6.3f} ms".format(
                dtype.__name__, runtime_min, runtime_avg, runtime_max
            )
        )
        result_row = [dtype.__name__] + [str(rt) for rt in all_runtimes]
        result_file.write(";".join(result_row) + "\n")


if __name__ == "__main__":
    if len(sys.argv) != 2:
        print("ERROR: Expected argument N.")
        sys.exit(1)
    else:
        N = int(sys.argv[1])

    dtypes = [np.int16, np.int32, np.int64, np.float32, np.float64]

    benchmarks = [
        (bench_zeros, "zeros"),
        (bench_ones, "ones"),
        (bench_range, "range"),
        (bench_sum, "sum"),
        (bench_max, "max"),
    ]

    for benchmark, label in benchmarks:
        filename = "results/result_{}_pandas_{}.csv".format(N, label)
        with open(filename, "w") as result_file:
            run_benchmark_repeated(
                benchmark, label, N, 100, dtypes, result_file
            )
